File: ThirdYear/prabhat-04_PrabhatKeshari_2125it1086_3/Week_3/Subsets.cpp
class Solution {
public:

    void solve(vector<int>&nums,vector<int>res,vector<vector<int>>&ans,int i){
        if(i>=nums.size()){
            ans.push_back(res);
            return;
        }
        solve(nums,res,ans,i+1);
        res.push_back(nums[i]);
        solve(nums,res,ans,i+1);
    }

    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> ans;
        vector<int> res;
        solve(nums,res,ans,0);
        return ans;
    }
};

File: ThirdYear/harshi706_HarshiAgrawal_2125it1200_3/Week_3/subsets.java
class Solution {
    public void help(int nums[],int i,ArrayList<Integer> subset,List<List<Integer>> ans){
        if(i>=nums.length){
        ans.add(new ArrayList<>(subset));
        return;
        }
        subset.add(nums[i]);
        help(nums,i+1,subset,ans);
        subset.remove(subset.size()-1);
        help(nums,i+1,subset,ans);
        }
    
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>>ans=new ArrayList<>();
        ArrayList<Integer>subset=new ArrayList<>();
        help(nums,0,subset,ans);
        return ans;
        
    }
}